### ISSUE TITLE
XPath root node & ancestor selectors not working as expected

### ISSUE THREAD

**Issue by anulman** [mischov/meeseeks#42]
Using https://www.seriouseats.com/recipes/2010/09/sous-vide-101-duck-breast-recipe.html as an example...

### Root node selector
`Meeseeks.all(doc, xpath("/*[@itemscope]))` is returning all root nodes.

In this case, this is expected behaviour (ie the root `<html>` node declares an `itemscope` attr), but the root node is still returned even if you rm the `itemscope` attr.

### Ancestor selectors
`Meeseeks.all(doc, xpath("//*[@itemprop][not(ancestor::*[@itemscope][2])]"))` is throwing an `ArithmeticError`. Specifically:

```
** (ArithmeticError) bad argument in arithmetic expression
    (meeseeks) lib/meeseeks/selector/xpath/expr/helpers.ex:358: Meeseeks.Selector.XPath.Expr.Helpers.plus_one/1
    (meeseeks) lib/meeseeks/selector/xpath/expr/predicate.ex:15: Meeseeks.Selector.XPath.Expr.Predicate.eval/4
    (elixir) lib/enum.ex:2824: Enum.all_list/2
    (elixir) lib/enum.ex:2872: Enum.filter_list/2
    (meeseeks) lib/meeseeks/selector/xpath/expr/path.ex:29: anonymous fn/5 in Meeseeks.Selector.XPath.Expr.Path.next_step/4
    (elixir) lib/enum.ex:1899: Enum."-reduce/3-lists^foldl/2-0-"/3
    (meeseeks) lib/meeseeks/selector/xpath/expr/function.ex:319: Meeseeks.Selector.XPath.Expr.Function.eval/4
    (meeseeks) lib/meeseeks/selector/xpath/expr/predicate.ex:11: Meeseeks.Selector.XPath.Expr.Predicate.eval/4
```

In this case, I would expect Meeseeks to find 11 matching results, ie all DOM nodes with an `itemprop` attr not nested in a secondary `itemscope`d node.
> At the example URL, the `reviewCount` & `ratingValue` nodes are nested in the item declared by the `aggregateRating` node

**Comment by mischov** [comment]
@anulman Thanks for bringing this to my attention!

I haven't had a chance to look at root node stuff yet, but I should have a solution for the ancestor selectors bit.

The logic for running predicates in step expressions was all messed up- the context wasn't being updated correctly, and the nodes weren't being filtered by the predicates correctly. It worked right for simple cases, but once things got complex it broke (as you discovered).

**Comment by mischov** [comment]
The root node selector part was just a bug where I didn't copy the filters when transpiling an absolute `XPath.Expr.Path` to a `Selector.Root`.

I'll need a little time to write tests and stuff, but I should have a fix for these bugs out pretty soon.

**Comment by mischov** [comment]
@anulman Both problems should be resolved in v0.9.4.

Thank you again for bringing these problems to my attention.

I'm closing the issue but please reopen it if your problems are not resolved.
